README.md
# Displaygrid

Displaygrid is a javascript/html framework to display a
variable sized grid which elements can be freely defined 
by special functions.

## Synopsis

displaygrid.html?<param1=value1+param2=value2+...>

# Parameters

  parameter    | description
---------------| -----------
  gridwidth    | grid witdth in cells
  gridheight   | grid height in cells
  cellwidth    | cell width in pixels
  cellheight   | cell heigth in pixels
  borderwidth  | border width in pixels
  borderheight | border heigth in pixels

## Requirements

Displaygrid requires a decent browser supporting javascript and html5 canvas.

## Testing/Diagnostics

displaygrid should display a grid using the defined dimensions (see
Parameters) and fill it with random colors. Input Parameters are
checked for positive integers but not for upper limits (so come on,
check it out and have your browser take a 4k grid using a 1 pixel cell).

Tested on Firefox 60+. No browwser capability checks are done, if it 
doesn't work for you , use another browser.

## Copyright
Displaygrid is copyright (c) 2018 ths.one

## License:
This Application and all its containing files is free software; you 
can redistribute it and/or modify it under the terms of the GNU 
General Public License as published by the Free Software Foundation; 
either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

## Known Bugs:
There probably are some, we'll let you know as soon as we have found them.

## Authors:
ths@ths.one




doc/changelog.txt
2018/11/29 - version 0.5.5 - ths
  - development start


common/jss/queryObj.js
//
// queryObj.js  - query parameter handling object
// copyright 2018 ths@ths.one
//
// created : 2018/12/01
// changed : 2018/12/02
// version : 0.5.0
//
// This Application and all its containing files is free software; you 
// can redistribute it and/or modify it under the terms of the GNU 
// General Public License as published by the Free Software Foundation; 
// either version 2 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but 
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

// common tools, not limited for queryObj
//

// function decodeQuery( qstr )
//   convert encoded chars & spaces
//   special chars are coded as %XX where XX stands for a 2 digit hexadecimal string
//   space are code as "+"

function decodeQuery( qstr )
{
  var i,c,qlen,rstr;

  qstr=qstr.toString();
  qlen = qstr.length;
  rstr = "";
  if (qlen==0) return rstr;

  i = 0;
  while(i<qlen)
  {
    c=qstr.charAt(i);
    switch(c)
    {
      case '+' :
        c=' '; 
        break;

      case '%' : 
        if (qstr.charAt(i+1) == '%')
        {
          i++;
          break;
        }
        if ( i<(qlen-2) )
        {
          c=String.fromCharCode(parseInt(qstr.substr(i+1,2),16));
          i+=2;
        }
        break;
    }
    i++;
    rstr += c;
  }
  return rstr;
}

// function encodeQuery( qstr )
//   convert everything not in a..zA..Z0..9_ to %XX hexadecimal format
//   and convert spaces to "+"

function encodeQuery( qstr )
{
  var i,cc,qlen,rstr;
  
  qstr=qstr.toString();
  qlen = qstr.length;
  rstr = "";
  if (qlen==0) return rstr;
  for (i=0;i<qlen;i++)
  {
    cc=qstr.charAt(i);
    if (cc==" ") cc="+";
    else  cc=escape(cc);
    rstr+=cc;
  }
  return rstr;
}

// query object definition
//

function queryObj( href, filter ) 
{

  // object properties

  this.url     = "";                // complete url without query string
  this.query   = "";                // complete query string 
  this.qlen    = 0;                 // query length

  this.pcount  = 0;                 // number of parameter pairs (name=value) in query
  this.pnames  = "";                // will contain parameter names
  this.pvalues = "";                // will contain parameter values
  
  this.filter  = filter;
  this.nvdiv   = "=";

  // object methods

  this.initObject  = QO_initObject;
  this.parseQuery  = QO_parseQuery;
  this.getValue    = QO_getValue;
  this.setValue    = QO_setValue;
  this.createQuery = QO_createQuery;

  // object setup
  
  this.initObject( href , filter );
}

function QO_initObject( href , filter )
{
  var hlen,qpos,qstr;
  
  this.url     = href;
  this.filter  = filter;
  
  qstr="";
  
  if (href) if ( (hlen=href.length) > 0)
  {
    // split full href into url and query-string
    qpos = href.indexOf("?");
    if ( (qpos>=0) && (qpos<hlen) )
    {
      if (qpos==0) this.url = "" ;
      else this.url=href.substring(0,qpos);
      qstr=href.substring(qpos+1,hlen) ;
    }
  }
  this.parseQuery(qstr);
  return this.pcount;
}

function QO_parseQuery( qstr )
{
  var p,parr,pc,px,pn,pv;
  
  this.query   = qstr
  this.qlen    = qstr.length;
  this.pstr    = "";
  this.plen    = 0; 
  this.pcount  = 0;
  this.pnames  = new Array();
  this.pvalues = new Array();
  
  if (!this.filter) this.filter="";
  pp=0;

  if (qstr) if (qstr.length > 0)
  {
    // split query into single parameters
    parr = qstr.split("&");
    pc   =  (parr) ?  parr.length : 0;

    for (p=0;p<pc;p++)
    {
      px = parr[p].split(this.nvdiv);
      pn = (px[0]) ? decodeQuery(px[0]) : decodeQuery(parr[p]) ;
      pv = (px[1]) ? decodeQuery(px[1]) : "" ;
      
      // allow only valid names if filter has been defined
      if ( (this.filter.length==0) || (this.filter.indexOf(pn)>=0) )
      {
        this.pnames[this.pcount]  = pn;
        this.pvalues[this.pcount] = pv;
        this.pcount++;
      }
    }  

  }
  return this.pcount;
}

function QO_getValue( pname )
{
  var p;
  for (p=0;p<this.pcount;p++)
  {
    if (this.pnames[p] == pname ) return this.pvalues[p];
  }
  return "";
}

function QO_setValue( pname , pval )
{
  var ov="";
  var p;
  for (p=0;p<this.pcount;p++)
  {
    if (this.pnames[p] == pname ) 
    {
      ov=this.pvalues[p];
      this.pvalues[p]=pval;
      return ov;
    }
  }
  this.pnames[p]  = pname;
  this.pvalues[p] = pval;
  this.pcount++;
  return ov;
}

function QO_createQuery( qupdate )
{
  var p;
  var qstr="";
  for (p=0;p<this.pcount;p++)
  {
    if (p>0) qstr+="&";
    qstr=qstr + encodeQuery(this.pnames[p]) + this.nvdiv + encodeQuery(this.pvalues[p]);
  }
  if (qupdate>0)
  {
    this.query=qstr;
    this.qlen=qstr.length;
  }
  return qstr;
}


common/jss/browserObj.js
//
// browserObj.js  - simple browser features request object
// copyright 2018 ths@ths.one
//
// created : 2018/12/01
// changed : 2018/12/02
// version : 0.5.0
//
// This Application and all its containing files is free software; you 
// can redistribute it and/or modify it under the terms of the GNU 
// General Public License as published by the Free Software Foundation; 
// either version 2 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but 
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

// TODO : update/test for latest browsers (can't stick to the 90's forever)

function browserObj( smode )
{
  // object properties

  this.infoStr           = "UNKNOWN";
  this.name              = "UNKNOWN";
  this.version           = 0.0;
  this.type              = "UNKNOWN";
  this.objModel          = "UNKNOWN";
  this.cssEnabled        = false;
  this.canvasSupported   = false;
  this.canvas2DSupported = false;
  this.debugMode         = 0;

  // object methods

  this.initObject    = BO_initObject;
  this.alertBrowser  = BO_alertBrowser;
  this.writeCSS      = BO_writeCSS;
  this.testFeatures  = BO_testFeatures;

  // object setup

  if (smode=="setup") this.initObject();
}

function BO_initObject( mode )
{
  this.infoStr  = navigator.userAgent;
  this.name     = navigator.appName;
  this.version  = parseFloat( navigator.appVersion );

  this.type     = "UNKNOWN";

  if (mode == "DEBUG") this.debugMode = 1 ;

  if ( this.infoStr.indexOf("MSIE") != -1)
    this.type="MSIE";
  else if ( this.infoStr.lastIndexOf("Gecko")  != -1 )
    this.type="MOZILLA";
  else if ( this.infoStr.lastIndexOf("Opera") != -1 )
    this.type="OPERA";
  else if ( this.infoStr.lastIndexOf("Konqueror") != -1 )
    this.type="KONQUEROR";
  else if ( this.infoStr.lastIndexOf("Mozilla")  != -1 )
    this.type="NETSCAPE";

  if ( document.documentElement ) this.objModel = "W3C";
  else if (document.all) this.objModel = "MSIE";
  else if (document.layers) this.objModel = "NETSCAPE";
  else this.objModel = "UNKNOWN";

  this.cssEnabled    = false;
  this.canvasEnabled = false;

  this.testFeatures() ;

  if (this.debugMode>2) this.alertBrowser();
}

function BO_alertBrowser()
{
  alert( "alertBrowser(): \nname = " + this.name +
         " \nVersion = " + this.version  +
         " \nType = " + this.type +
         " \nObjectModel = " + this.objModel +
         " \ncssEnabled = " + this.cssEnabled +
         " \ncanvasSupported = " + this.canvasSupported +
         " \ncanvas2DSupported = " + this.canvas2DSupported +
         " \nstring  = " + this.infoStr
       );
}

function BO_writeCSS()
{
  document.write( "<style type=\"text/css\"><!--\n" );
  document.write( "#testlayer{position: absolute; top: 50px; left: 0px;}\n" );
  document.write( "--></style>\n" );
  document.write( "<div id=\"testlayer\" style=\"position: absolute; top: 50px; left: 0px;\">\n  gagagag &nbsp;\n</div>\n");
}

function BO_testFeatures()
{
  var tobj ;
  var tpos = 0;

  if (this.objModel == "NETSCAPE")
    tpos = ( tobj = document.layers['testlayer'] ) ? parseInt( tobj.top) : 0;
  else if (this.objModel == "MSIE")
    tpos = ( tobj=document.all['testlayer'] ) ? parseInt( tobj.style.top ) : 0;
  else if (this.objModel == "W3C")
  {
    tpos = ( tobj=document.getElementById('testlayer')) ? parseInt( tobj.style.top ) : 0;
    this.canvasSupported   = !!window.HTMLCanvasElement;
    this.canvas2DSupported = !!window.CanvasRenderingContext2D;
  }

  this.cssEnabled = (tpos==50) ? true : false;
  if ( this.debugMode > 2 ) alert( "BO_testCSS(): testlayer object = " + String(tobj) ) ;
}



common/jss/displaygrid.js
//
// displaygrid.js  - javascript source for diplaygrid
// copyright 2018 ths@ths.one
//
// created : 2018/12/01
// changed : 2018/12/02
// version : 0.5.0
//
// This Application and all its containing files is free software; you 
// can redistribute it and/or modify it under the terms of the GNU 
// General Public License as published by the Free Software Foundation; 
// either version 2 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but 
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//



common/jss/globalObj.js
//
// globalObj.js  - global object for handling global settings
// copyright 2018 ths@ths.one
//
// created : 2018/12/01
// changed : 2018/12/03
// version : 0.5.0
//
// This Application and all its containing files is free software; you 
// can redistribute it and/or modify it under the terms of the GNU 
// General Public License as published by the Free Software Foundation; 
// either version 2 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but 
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

// global defaults
//

DEFAULT_GRIDWIDTH    = 200 ;
DEFAULT_GRIDHEIGHT   = 200 ;
DEFAULT_CELLWIDTH    = 10  ;
DEFAULT_CELLHEIGHT   = 10  ;
DEFAULT_BORDERWIDTH  = 2   ;
DEFAULT_BORDERHEIGHT = 2   ;

// global configuration object

function globalObj( smode )
{
  // status properties
  
  this.objQuery     = new queryObj("","");

  this.rootMode    = -1;
  this.testMode    = -1;
  this.debugMode   = -1;
  this.action      = "";
  this.search      = "";
  this.language    = "";

  this.gridWidth    = DEFAULT_GRIDWIDTH ;
  this.gridHeight   = DEFAULT_GRIDHEIGHT ;
  this.cellWidth    = DEFAULT_CELLWIDTH ;
  this.cellHeight   = DEFAULT_CELLHEIGHT ;
  this.borderWidth  = DEFAULT_BORDERWIDTH ;
  this.borderHeight = DEFAULT_BORDERHEIGHT ;

  // object methods

  this.initObject    = GO_initObject;
  this.updateQuery   = GO_updateQuery;
  this.updateValues  = GO_updateValues;
  this.callValues    = GO_callValues;
  this.alertValues   = GO_alertValues;

  // object setup

  if (smode=="setup") this.initObject();
}

function GO_initObject() 
{
  var pc;
  // read query string first
  pc=this.objQuery.initObject(self.location.href,"ROOT:DEBUG:TEST:ACTION:gridwidth:gridheight:cellwidth:cellheight:borderwidth:borderheight"); 

  // will be used only if root=GAGA is set, otherwise we use top.name
  if ( this.objQuery.getValue("ROOT") == "GAGA" ) this.rootMode  = 1;
  else this.rootMode  = 0;

  if (pc>0) 
  {
    this.updateQuery();
    this.qmethod="GET";
  }
  else
  {
    this.qmethod="JS";
    this.updateValues();
  }
}  

function GO_updateQuery() 
{
  var query      = this.objQuery.createQuery(1);
  this.updateValues();
}

function GO_updateValues()
{

  this.testMode    = parseInt(this.objQuery.getValue("TEST"));
  if (isNaN(this.testMode)) this.testMode=0 ;
  this.debugMode   = parseInt(this.objQuery.getValue("DEBUG")) ;
  if (isNaN(this.debugMode)) this.debugMode=0 ;

  this.action      = this.objQuery.getValue("ACTION");
  this.language    = this.objQuery.getValue("LANGUAGE");

  this.gridWidth       = parseInt(this.objQuery.getValue("gridwidth")) ;
  if ( (isNaN( this.gridWidth )) || ( this.gridWidth <= 0 ) ) this.gridWidth = DEFAULT_GRIDWIDTH ;
  this.gridHeight      = parseInt(this.objQuery.getValue("gridheight")) ;
  if ( (isNaN( this.gridHeight )) || ( this.gridHeight <= 0 ) ) this.gridHeight = DEFAULT_GRIDHEIGHT ;

  this.cellWidth       = parseInt(this.objQuery.getValue("cellwidth")) ;
  if ( (isNaN( this.cellWidth )) || ( this.width <= 0 ) ) this.cellWidth = DEFAULT_CELLWIDTH ;
  this.cellHeight      = parseInt(this.objQuery.getValue("cellheight")) ;
  if ( (isNaN( this.cellHeight )) || ( this.cellHeight <= 0 ) ) this.cellHeight = DEFAULT_CELLHEIGHT ;

  this.borderWidth     = parseInt(this.objQuery.getValue("borderwidth")) ;
  if ( (isNaN(this.borderWidth )) || ( this.width <= 0 ) ) this.borderWidth = DEFAULT_BORDERWIDTH ;
  this.borderHeight    = parseInt(this.objQuery.getValue("borderheight")) ;
  if ( (isNaN( this.borderHeight )) || ( this.borderHeight <= 0 ) ) this.borderHeight = DEFAULT_BORDERHEIGHT ;

}

function GO_callValues()
{
  return "?rootMode=" + this.rootMode +
         "+testMode=" + this.testMode  +
         "+debugMode=" + this.debugMode +
         "+action=" + this.action +
         "+gridWidth=" + this.gridWidth  +
         "+gridHeigth=" + this.gridHeight +
         "+cellWidth=" + this.cellWidth  +
         "+cellHeigth=" + this.cellHeight +
         "+borderWidth=" + this.borderWidth  +
         "+borderHeigth=" + this.borderHeight ;
}

function GO_alertValues()
{
  alert( "GO_alertValues(): \nrootMode = " + this.rootMode +
         " \ntestMode     = " + this.testMode  +
         " \ndebugMode    = " + this.debugMode +
         " \naction       = " + this.action +
         " \ngridWidth    = " + this.gridWidth  +
         " \ngridHeigth   = " + this.gridHeight +
         " \ncellWidth    = " + this.cellWidth  +
         " \ncellHeigth   = " + this.cellHeight +
         " \nborderWidth  = " + this.borderWidth  +
         " \nborderHeigth = " + this.borderHeight
       );
}


